### PR TITLE
Weighted inner product

### PR DESCRIPTION
**Pull request by ezander**
Hi Lionel,
here's the pull request for the weighted inner product function, with unit tests and changes to get_dL as proposed in the forum.

### CODE REVIEW

File: core/VBA_get_dL.m [hunk 1]
```diff
@@ -47,15 +47,15 @@
         vy = gx.*(1-gx) ;
         ddydphi = dG_dPhi*(dy./vy);
         temp = y./(gx).^2 - (y-1)./(1-gx).^2;
-        d2gdx2 = dG_dPhi*diag(temp)*dG_dPhi';    
+        d2gdx2 = VBA_weighted_inner_product(dG_dPhi, [], temp);
```
Comment (lionel-rigoux) [inline core/VBA_get_dL.m]: Dimensions of d2gdx2 are wrong in the new code. Should it be?
```suggestion
        d2gdx2 = VBA_weighted_inner_product(dG_dPhi', [], temp);
```

File: core/VBA_get_dL.m [hunk 2]
```diff
@@ -47,15 +47,15 @@
...
         logL = y'*log(gx) + (1-y)'*log(1-gx);
         dy2 = sum(temp);
         
    case 2   %--- multinomial       
         gx = VBA_finiteBinomial (gx);
         vy = gx.*(1-gx) ;
         ddydphi = dG_dPhi*(y./gx);
-        d2gdx2 = dG_dPhi*diag(y./gx.^2)*dG_dPhi';
+        d2gdx2 = VBA_weighted_inner_product(dG_dPhi, [], y./gx.^2);
```
Comment (lionel-rigoux) [inline core/VBA_get_dL.m]: Same comment as for binomial case

**Comment by lionel-rigoux** [comment]
Hi,
I gave a try to your code. I have two issues with your suggestion:
- There is a problem of dimensions in the new code. I guess this is just a missing transpose (see comments in the code).
- For the common user case, the dimension of the observation is rather small (eg one binary choice). In that case, your proposed solution is way slower than the current code. Therefore, I do not see the bsxfun solution as a generic workaround and I wouldn't approve the pull request as it is because of this.

What are the dimensions of your current problem? If you give me more details I may be able to help you find a solution that will cover your case without slowing the toolbox in the general case.
Best

**Comment by ezander** [comment]
Hi Lionel,

1. Yes, you're correct here - there is a missing transpose.
2. I have usually some hundreds to some thousands of of observations (they come from a discrete choice model with approx. a few hundred or thousand respondents and 10 to 20 choice situations per respondent). So, e.g. in the case I tested the toolbox `dG_dPhi` had dimensions 8x50000 and in this case the `bsxfun` is way faster.

**Comment by lionel-rigoux** [comment]
I'm not sure I am following.
What is the dimension of your `y`matrix? Is it splitted into multiple sources?
Do you aggregate multiple subjects?
How many parameters do you have?

**Comment by ezander** [comment]
I assume you mean with `y` the matrix of observations... I my case it's a boolean array of size `(N,T,J)` where `N` is the number of respondents (subjects) in the range of 100-1000, `T` the number of choice situations in the range 10-20, and `J` the number of choices per choice situation around 3-5. So, the total number of observations is somewhere between 3000 and 100000.
Further, I have `K` attributes per choice (explanatory variables, around 5-50). The individual parameters I want to infer have thus size `K*N` plus some global parameters of `O(K)`.
If I understand you correctly, I'm neither aggregating nor splitting anything here.
The parameters I have, are

**Comment by lionel-rigoux** [comment]
Ok, I am getting even more confused now.

The observation matrix should be two dimensional. Each column represent an independent measurement (timepoint), multiple lines are for concomitent measurments (ie. with potential correlation structure). Are you using a 3D matrix? I am suprised this might work at all.

If you have multiple binary observations (multiple lines in the matrix y), you could split them in multiple sources (see [here](http://mbb-team.github.io/VBA-toolbox/wiki/Multisources/#specifying-the-observation-mixture)) to mitigate the memory issue related to the `diag` in VBA_get_dL.

Also, it seems that you fit one set of parameters for each subject. Why don't you fit each subject separately, and then perform a group analysis on the respective posteriors? If you only have a subset of shared parameters, something like a mixed effect approach could be more efficient here (look for MFX functions and wiki entries).

**Comment by lionel-rigoux** [comment]
Did you solve your performance issue?

**Comment by lionel-rigoux** [comment]
Hello,
Is there any update on this issue? Have you solved you performance problem?
If I don't get any feedbacks I will close the issue and the pull request.

**Pull request closed without merging**